app/controllers/fhir/r4/api_controller.rb

```diff
@@ -342,12 +335,100 @@ def create
                             created_by: @current_actor_label,
                             comment: "New vaccine added via API (ID: #{resource.id}).")
       end
+    when 'observation'
+      return if doorkeeper_authorize!(*OBSERVATION_WRITE_SCOPES)
+
+      resource, fhir_map = build_laboratory(contents)
+
+      ActiveRecord::Base.transaction do
+        unless referenced_patient_valid_for_client?(resource, :patient_id) && resource.save(context: :api) && fhir_map.all? { |_k, v| v[:errors].blank? }
+          req_json = JSON.parse(request.body.string)
+          status_unprocessable_entity(resource, fhir_map, req_json) && (raise SaveError)
+        end
+
+        Rails.logger.info "Created Lab Result (ID: #{resource.id}) for Patient with ID: #{resource.patient_id}"
+        History.lab_result(patient: resource.patient_id,
+                           created_by: @current_actor_label,
+                           comment: "New lab result added via API (ID: #{resource.id}).")
+      end
     else
       status_not_found && return
     end
     status_created(resource.as_fhir) && return
   rescue JSON::ParserError
     status_bad_request(['Invalid JSON in request body'])
+  rescue SaveError
+    nil
+  end
+
+  # Create a set of resources as an atomic action
+  #
+  # Supports (writing): Patient, Observation
+  #
+  # POST /fhir/r4
+  def transaction
+    status_unsupported_media_type && return unless content_type_header?('application/fhir+json')
+
+    # Must have both Observation and Patient write scopes, since both can be written
+    return if doorkeeper_authorize!(*OBSERVATION_WRITE_SCOPES)
+    return if doorkeeper_authorize!(*PATIENT_WRITE_SCOPES)
```

Comment:
Actually @ngfreiter since we are deciding to allow bulk enrollment here, the authorization needs to be modified. A bundle could only have Patients, so here you'd only want to check for patient write scope. And then I think you'd have to add a step in [this loop](https://github.com/SaraAlert/SaraAlert/pull/870/files#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R753) to check for Observation write scopes in the event that there is at least one Observation.

Reply:
Refactored in 3fd3d89.